flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1495,40 +1524,65 @@ class ScalarExprLowering {
     auto idxTy = builder.getIndexType();
     llvm::SmallVector<mlir::Value, 8> exprShape;
+    auto loc = getLoc();
     if (arrTy.hasConstantShape()) {
       for (auto s : arrShape)
-        exprShape.push_back(builder.createIntegerConstant(getLoc(), idxTy, s));
+        exprShape.push_back(builder.createIntegerConstant(loc, idxTy, s));
     } else {
       for (auto s : temp.getShapeOperands())
-        exprShape.push_back(builder.createConvert(getLoc(), idxTy, s));
+        exprShape.push_back(builder.createConvert(loc, idxTy, s));
     }
-    auto shape = builder.create<fir::ShapeOp>(getLoc(), shapeTy, exprShape);
+    auto shape = builder.create<fir::ShapeOp>(loc, shapeTy, exprShape);
     mlir::Value slice;
-    auto arrLd = builder.create<fir::ArrayLoadOp>(getLoc(), arrTy, temp, shape,
+    auto arrLd = builder.create<fir::ArrayLoadOp>(loc, arrTy, temp, shape,
                                                   slice, llvm::None);
     auto loopRes = Fortran::lower::createSomeNewArrayValue(
-        converter, arrLd,
-        Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(x)), symMap,
-        stmtCtx);
+        converter, arrLd, {}, toEvExpr(x), symMap, stmtCtx);
     auto tempRes = temp.getResult();
-    builder.create<fir::ArrayMergeStoreOp>(getLoc(), arrLd,
-                                           fir::getBase(loopRes), tempRes);
+    builder.create<fir::ArrayMergeStoreOp>(loc, arrLd, fir::getBase(loopRes),
+                                           tempRes);
     return tempRes;
   }
 
+  /// Lower an array value as an argument. This argument can be passed as a box
+  /// value, so it may be possible to avoid making a temporary.
+  template <typename A>
+  fir::ExtendedValue asArrayArg(const Fortran::evaluate::Expr<A> &x) {
+    return std::visit([&](const auto &e) { return asArrayArg(e, x); }, x.u);
+  }
+  template <typename A, typename B>
+  fir::ExtendedValue asArrayArg(const Fortran::evaluate::Expr<A> &x,
+                                const B &y) {
+    return std::visit([&](const auto &e) { return asArrayArg(e, y); }, x.u);
+  }
+  template <typename A, typename B>
+  fir::ExtendedValue asArrayArg(const Fortran::evaluate::Designator<A> &,
+                                const B &x) {
+    // Designator is being passed as an argument to a procedure. Lower the
+    // expression to a boxed value.
+    return Fortran::lower::createSomeArrayBox(converter, toEvExpr(x), symMap,
+                                              stmtCtx);
+  }
+  template <typename A, typename B>
+  fir::ExtendedValue asArrayArg(const A &, const B &x) {
+    // If the expression to pass as an argument is not a designator, then create
+    // an array temp.
+    return asArray(x);
+  }
+
   template <typename A>
   fir::ExtendedValue gen(const Fortran::evaluate::Expr<A> &x) {
     if (isScalar(x) || Fortran::evaluate::UnwrapWholeSymbolDataRef(x))
       return std::visit([&](const auto &e) { return genref(e); }, x.u);
+    if (useBoxArg)
```

Comment:
I have a bit of a bad feeling that we could find a corner case where we can re-enter this when lowering a very non trivial designator, but I could not find an example yet, so that may be just OK.
Should it be set to false temporarily around the `asArrayArg` call ?

Reply:
I don't think that's possible from the structure of the Ev::Expr data structures.